Action management modules and specification

MTUC08 opened the pull request:
This pull request contains our action specification and basic implementation of our action modules. We have split all our supported actions into five types: action_item, action_direction, action_npc, action_item_npc, and action_item_item, and begun implementation with structs and functions we assume will be implemented by game state. This is the culmination of the first two sprints.

borjasotomayor commented:
@MTUC08: Please read https://uchicago-cs.github.io/cmsc22000/projects/2019/spec/#pull-requests-to-master. More specifically: 

> If you want to merge code into master, your code must be reviewed by (1) someone in your team, (2) someone in another team, and (3) your senior developer. 

Please adjust the list of reviewers accordingly (and remember that your senior developer is @mpcsh)

borjasotomayor commented on .travis.yml:
```diff
@@ -13,4 +13,4 @@ language: c
 
 script:
-  - make
+  - make -C 
```
This line breaks the build. You need to add a `Makefile` in `src/action_management/` and then update the `Makefile` at the root of the repository so it will call your Makefile. There should be no need for you to modify the `.travis.yml` file.

MTUC08 replied:
Hi, sorry about that, I’m not sure why that edit is showing up, we’ll implement a makefile asap

mpcsh commented:
@MTUC08 following up on Borja's comment, it looks like you've requested reviews from your whole team, which is not required. That will slow down the review process significantly. Instead, you should ask for one teammate to review. Also, please edit the title of this PR to be more descriptive (what files are being merged? what do they do?).

@yoonsjeong if you're going to leave a review, it can't just be comments - you need to use the "Add your review" button at the top of the page. For now, you'll want to select "Request changes".

SnowballAntrobus commented on src/action_management/include/structs.h:
```diff
@@ -0,0 +1,42 @@
+/*  A room struct containing:
+ * - NPCs: linked list of pointers to NPC_structs
+ * - objects: linked list of pointers to object_structs
+ * - overall_map: a 2D array that would allow us to use x y coordinates
```
maybe a graph would be better

MTUC08 replied:
This structs.h is outdated I think, we're just going to go off of game state structs

MTUC08 replied:
just updated game.h and we don't really need to know what the map is yet, so we leave that to game state

MTUC08 commented:
Update: I moved all of the action structs into a new file titled action_structs.h that I included in actionmanagement.h and the newly titled and updated game.h file, previously named structs.h.The new game.h file includes all of the struct components and game state functions we needed to assume existed in game state in order to implement our functions. I then added an additional check within actionmanagement.c for all functions involving actions with items to ensure that the action can be applied to the action, and edited coding errors within the functions. In order to make sure everything compiled, I forward declared a list_t struct (although we will be implementing that in later sprints) and added the get_action_kinds1-5 functions to the actionmanagement.h file (they will also be implemented in later sprints due to the reliance on a list_t struct). I additionally changed the kind component to an enum rather than an enum * as I don't think the pointer is actually necessary since we do not have a function that needs to create the enum. I also added skeleton functions for our assumed game functions and the get functions so that our code would compile.

MTUC08 commented on src/action_management/include/actionmanagement.h:
```diff
@@ -0,0 +1,239 @@
+ * - 0 on success
+ * - 1 on failure
+ */
+int action_item(game_t *g, action_t *a, item_t *i);
```
we reflected the word because when the other teams are dealing with our functions, action_item would be much more intuitive than action_kind1, which they would have to look up.

MTUC08 commented on src/action_management/include/actionmanagement.h:
```diff
@@ -0,0 +1,239 @@
+
+
+/* 
+ * A function that returns the synonyms of an action as a list of enums 
```
I changed it to strings instead. Thanks!

a-matai commented on docs/actions.md:
```diff
@@ -0,0 +1,335 @@
+**Effects**:<br>
+When an action is completed successful, there are multiple types of effects<br>
+that can occur, depending on which action is committed on a certain item.<br>
+
```
I really like these specific definitions, though I think that it might be better to provide an example/input/output—modeling the flow can help understand the context? Though it's fine because it is clarified after

MTUC08 replied:
@lucyyutingli @SnowballAntrobus

lucyyutingli replied:
@a-matai What do you mean by examples? Do you mean some similar to those we listed under the Kinds of actions at the top?

SnowballAntrobus replied:
https://github.com/uchicago-cs/chiventure/issues/126

a-matai commented on docs/actions.md:
```diff
@@ -0,0 +1,335 @@
+
+## Items and NPC's:
+
+These are the general conditions and requirements for NPC's.
```
It says for NPCs but goes into items? Just needs to be clarified I think?

MTUC08 replied:
@lucyyutingli @SnowballAntrobus

SnowballAntrobus replied:
https://github.com/uchicago-cs/chiventure/issues/126

lucyyutingli replied:
Just fixed this in the issue Dante mentions.

a-matai commented on docs/actions.md:
```diff
@@ -0,0 +1,335 @@
+- Player state may be changed
+
+
+## Actions:
```
The spec below is great and really thorough. There are just two things I'm thinking of:

Instead of K1, k2, etc given that the text is long, perhaps something that reminds the reader—a more specific name like INTERACT or MOVE or SPEAK or TAKE might help? 
I also think that now that game-state has the data structs out (sorry bout that), it might be good to refactor those in because action management will be dealing with those from now on?

MTUC08 replied:
@lucyyutingli @SnowballAntrobus

SnowballAntrobus replied:
https://github.com/uchicago-cs/chiventure/issues/126

a-matai commented on src/action_management/include/action_structs.h:
```diff
@@ -0,0 +1,75 @@
+    GO,      // synonyms: "walk"
+    ENTER,
+
+/* KIND 3 ACTIONS */
```
same here: perhaps give them a name instead of a number? easier to follow "K2 -direction types"

MTUC08 replied:
That's more for our reference, but I can add it

a-matai commented on src/action_management/include/action_structs.h:
```diff
@@ -0,0 +1,75 @@
+ * - synonyms: the synonyms that would also be allowed to call action
+ * - parameters: an order-sensitive linked list of object_type enums
+*/
+typedef struct {
```
this is cool

a-matai commented on src/action_management/include/actionmanagement.h:
```diff
@@ -0,0 +1,198 @@
+
+/*
+ * A function that initializes an action
+ * Will determine the action_type, depending on what action is given
```
I might be bad but it's unclear what action_type is: is it a kind or action_t?

MTUC08 replied:
removed that line as we will be passing the kind to the function rather than determining it internally

a-matai commented on src/action_management/include/actionmanagement.h:
```diff
@@ -0,0 +1,198 @@
+ * - 0 on success
+ * - 1 on failure
+ */
+int action_direction(game_t *g, action_t *a, direction_t *d);
```
I'm a little confused by the direction kind—aren't they all just the same action with the same effect?

on this—I see that you have fwd declaration. Would it be worth integrating game-state's path struct into this?

mpcsh replied:
@a-matai to your first point, no, they're not all the same. For example, if you're standing in a room with a locked door to the east and an open door to the north, then going east and north will have different conditions and effects.

To your second point, yes, this will be integrated, but I think that will happen as a fast-follow rather than as part of this PR.

a-matai commented on src/action_management/include/actionmanagement.h:
```diff
@@ -0,0 +1,198 @@
+ * - 0 on success
+ * - 1 on failure
+ */
+int action_npc(game_t *g, action_t *a, npc_t *n);
```
how do you measure the response with NPCs—do we support multiple interactions or just one? 

I think this might be more relevant to WDL though I'm not sure.

MTUC08 replied:
I think we are currently in talks with the other teams in general to determine if npc's will be implemented at all. This may be removed in a later sprint

a-matai commented on src/action_management/include/actionmanagement.h:
```diff
@@ -0,0 +1,198 @@
+ */
+int action_item_item(game_t *g, action_t *a, item_t *direct, item_t *indirect);
+
+
```
this looks solid. I think we would all need to meet to discuss what is "returned" by the actual game when an interaction occurs as well.

MTUC08 replied:
Yeah we can talk about that for this or the next sprint

a-matai commented on src/action_management/include/game.h:
```diff
@@ -0,0 +1,112 @@
+#include "action_structs.h"
+
+
+/* In the absence of the Game State team's code, 
```
lmao nice

a-matai commented on src/action_management/include/game.h:
```diff
@@ -0,0 +1,112 @@
+/* A NPC struct containing:
+ * - dialogue: a line of dialogue NPC will say when spoken to 
+ */
+typedef struct {
```
@liuyifan-eric @lindsey-hsi @patrickkoopmans @jessxia once we get integrated and tested we'll need to implement these

MTUC08 replied:
The npc struct may be irrelevant if the class decides to remove npc support

a-matai commented on src/action_management/include/game.h:
```diff
@@ -0,0 +1,112 @@
+ * - status_change: a string notifying the player of the status change
+ */
+typedef struct {
+  int *allowed_actions;
```
given that WDL will be storing these in our structs when the game is initialized, we need to figure out how to make our item structs compatible and accessible to yours, vice-versa.

I think the largest difficulty is in sorting them to the array of allowed_action enums, but you have functions for that so I assume we'll figure that out this sprint. 

The other question I have is maybe just unclear from documentation, but if I have an apple, I can eat or throw it, right? One could affect my health, the other wouldn't? How would we sort between which actions affect the person's health or don't?

MTUC08 replied:
Our functions handle for that. For ex: consume will affect the person's health, but push will not

mpcsh commented on src/action_management/include/common.h:
```diff
@@ -0,0 +1,7 @@
+#ifndef _COMMON_H_
+#define _COMMON_H_
+
+#define SUCCESS (0)
```
This file and these macros should probably be moved to the top-level `/include` directory - every team can benefit from and use this.

mpcsh commented on src/action_management/src/actionmanagement.c:
```diff
@@ -0,0 +1,346 @@
+    action_t *a = malloc(sizeof(action_t));
+
+    if (a == NULL)
+    {
```
This brace style is not consistent with the other files in this PR. Please run all your code through `astyle`, which is installed on the CSIL machines and on the VM, and use the Stroustrup style. You can read more by running `man astyle` in the terminal.

mpcsh commented on src/action_management/include/actionmanagement.h:
```diff
@@ -0,0 +1,198 @@
+ * Returns:
+ * - the canonical name of the action as a string
+*/
+char *get_action_cname(action_t *a);
```
This function and the other two getters below are not useful - they should just be removed. Users of the `action_t` struct will just access the fields directly (e.g. `action->c_name`).

mpcsh commented on src/action_management/src/actionmanagement.c:
```diff
@@ -0,0 +1,346 @@
+ * along with the implementation/inclusion of a list_t struct */
+list_t *get_actions_kind1() {
+    return NULL;
+}
```
Even though these are dummy functions, you need two line breaks between all functions.

mpcsh commented on src/action_management/src/actionmanagement.c:
```diff
@@ -0,0 +1,346 @@
+    case ITEM:
+        actions = get_actions_kind1();
+        return actions;
+	    break;
```
`break` statements after `return` are redundant and unreachable, because you've already returned from the function. Please remove all such redundant `break`s.

SnowballAntrobus commented on docs/actions.md:
```diff
@@ -0,0 +1,296 @@
+﻿# Documentation of Action Specifications
+
+#### Definitions
+**Conditions**: <br>
```
```suggestion
**Conditions**:

```

SnowballAntrobus commented on docs/actions.md:
```diff
@@ -0,0 +1,296 @@
+must be met in order for an action to be successful. When the action is successful<br>
+there are potential effects that can occur. <br>
+
+**Effects**:<br>
```
```suggestion
**Effects**:

```

SnowballAntrobus commented on docs/actions.md:
```diff
@@ -0,0 +1,296 @@
+
+#### Definitions
+**Conditions**: <br>
+When an action is requested on an item, these conditions<br>
```
```suggestion
When an action is requested on an item, these conditions

```

SnowballAntrobus commented on docs/actions.md:
```diff
@@ -0,0 +1,296 @@
+#### Definitions
+**Conditions**: <br>
+When an action is requested on an item, these conditions<br>
+must be met in order for an action to be successful. When the action is successful<br>
```
```suggestion
must be met in order for an action to be successful. When the action is successful

```

SnowballAntrobus commented on docs/actions.md:
```diff
@@ -0,0 +1,296 @@
+**Conditions**: <br>
+When an action is requested on an item, these conditions<br>
+must be met in order for an action to be successful. When the action is successful<br>
+there are potential effects that can occur. <br>
```
```suggestion
there are potential effects that can occur. 

```

SnowballAntrobus commented on docs/actions.md:
```diff
@@ -0,0 +1,296 @@
+there are potential effects that can occur. <br>
+
+**Effects**:<br>
+When an action is completed successful, there are multiple types of effects<br>
```
```suggestion
When an action is completed successful, there are multiple types of effects

```

SnowballAntrobus commented on docs/actions.md:
```diff
@@ -0,0 +1,296 @@
+
+**Effects**:<br>
+When an action is completed successful, there are multiple types of effects<br>
+that can occur, depending on which action is committed on a certain item.<br>
```
```suggestion
that can occur, depending on which action is committed on a certain item.

```

SnowballAntrobus commented on docs/actions.md:
```diff
@@ -0,0 +1,296 @@
+
+
+## Kinds of Actions:
+Kind 1: ACTION `<ITEM>` <br>
```
```suggestion
Kind 1: ACTION `<ITEM>`

```

SnowballAntrobus commented on docs/actions.md:
```diff
@@ -0,0 +1,296 @@
+A **Kind 1** action is an action that makes the player do something to an item.<br>
+Examples: EAT apple, OPEN door, PUSH box, etc.<br>
+
+Kind 2: ACTION `<DIRECTION>`<br>
```
```suggestion
Kind 2: ACTION `<DIRECTION>`

```

SnowballAntrobus commented on docs/actions.md:
```diff
@@ -0,0 +1,296 @@
+Examples: EAT apple, OPEN door, PUSH box, etc.<br>
+
+Kind 2: ACTION `<DIRECTION>`<br>
+A **Kind 2** action is an action that makes the player walk in a certain direction, most likely into a new area/room.<br>
```
```suggestion
A **Kind 2** action is an action that makes the player walk in a certain direction, most likely into a new area/room.

```

SnowballAntrobus commented on docs/actions.md:
```diff
@@ -0,0 +1,296 @@
+
+Kind 2: ACTION `<DIRECTION>`<br>
+A **Kind 2** action is an action that makes the player walk in a certain direction, most likely into a new area/room.<br>
+Examples: GO west, WALK east, GO left, etc.<br>
```
```suggestion
Examples: go west, go left, etc.

```

SnowballAntrobus commented on docs/actions.md:
```diff
@@ -0,0 +1,296 @@
+A **Kind 2** action is an action that makes the player walk in a certain direction, most likely into a new area/room.<br>
+Examples: GO west, WALK east, GO left, etc.<br>
+
+Kind 3: ACTION `<ITEM> <ITEM>`<br>
```
```suggestion
Kind 3: ACTION `<ITEM>` ` <ITEM>`

```

SnowballAntrobus commented on docs/actions.md:
```diff
@@ -0,0 +1,296 @@
+Examples: GO west, WALK east, GO left, etc.<br>
+
+Kind 3: ACTION `<ITEM> <ITEM>`<br>
+A **Kind 3** action is an action that makes the player interact with two items. <br>
```
```suggestion
A **Kind 3** action is an action that makes the player interact with two items.

```

SnowballAntrobus commented on docs/actions.md:
```diff
@@ -0,0 +1,296 @@
+## Kinds of Actions:
+Kind 1: ACTION `<ITEM>` <br>
+A **Kind 1** action is an action that makes the player do something to an item.<br>
+Examples: EAT apple, OPEN door, PUSH box, etc.<br>
```
```suggestion
Examples: eat apple, open door, push box, etc.

```

SnowballAntrobus commented on docs/actions.md:
```diff
@@ -0,0 +1,296 @@
+
+## Kinds of Actions:
+Kind 1: ACTION `<ITEM>` <br>
+A **Kind 1** action is an action that makes the player do something to an item.<br>
```
```suggestion
A **Kind 1** action is an action that makes the player do something to an item.

```

SnowballAntrobus commented on docs/actions.md:
```diff
@@ -0,0 +1,296 @@
+
+Kind 3: ACTION `<ITEM> <ITEM>`<br>
+A **Kind 3** action is an action that makes the player interact with two items. <br>
+Examples: PUT apple on table, PUSH box on button, USE key on door, etc<br>
```
```suggestion
Examples: put apple on table, push box on button, use key on door, etc

```

SnowballAntrobus commented on docs/actions.md:
```diff
@@ -0,0 +1,296 @@
+
+## Actions:
+
+These are specific conditions and effects that are affiliated with certain actions. This means that along with the general<br>
```
```suggestion
These are specific conditions and effects that are affiliated with certain actions. This means that along with the general

```

SnowballAntrobus commented on docs/actions.md:
```diff
@@ -0,0 +1,296 @@
+## Actions:
+
+These are specific conditions and effects that are affiliated with certain actions. This means that along with the general<br>
+conditions and effects listed above, there are special conditions and effects tha can also be included with specific actions.<br>
```
```suggestion
conditions and effects listed above, there are special conditions and effects tha can also be included with specific actions.

```

SnowballAntrobus commented on docs/actions.md:
```diff
@@ -0,0 +1,296 @@
+
+These are specific conditions and effects that are affiliated with certain actions. This means that along with the general<br>
+conditions and effects listed above, there are special conditions and effects tha can also be included with specific actions.<br>
+If items have "No additional ..." listed, they follow the only follow the determined general conditions/requirements listed above, and <br>
```
```suggestion
If items have "No additional ..." listed, they follow the only follow the determined general conditions/requirements listed above, and

```

SnowballAntrobus commented on docs/actions.md:
```diff
@@ -0,0 +1,296 @@
+These are specific conditions and effects that are affiliated with certain actions. This means that along with the general<br>
+conditions and effects listed above, there are special conditions and effects tha can also be included with specific actions.<br>
+If items have "No additional ..." listed, they follow the only follow the determined general conditions/requirements listed above, and <br>
+have no excpetions.<br>
```
```suggestion
have no excpetions.

```

SnowballAntrobus commented on docs/actions.md:
```diff
@@ -0,0 +1,296 @@
+
+## Default:
+
+The action's behavior when called on an `<ITEM>` or [NPC] that does not support the action.
```
```suggestion
The action's behavior when called on an `<ITEM>` or `[NPC]` that does not support the action.
```

SnowballAntrobus commented on docs/actions.md:
```diff
@@ -0,0 +1,296 @@
+- A new room may be available
+
+### WDL:
+        - open:
```
```suggestion
```
        - open:
```
```

SnowballAntrobus commented on docs/actions.md:
```diff
@@ -0,0 +1,296 @@
+- A room may no longer be available 
+
+### WDL:
+         - close:
```
```suggestion
```
         - close:
```
```

SnowballAntrobus commented on docs/actions.md:
```diff
@@ -0,0 +1,296 @@
+- Long description of item will be printed
+
+### WDL:
+        - examine:
```
```suggestion
```
        - examine:
```
```

SnowballAntrobus commented on docs/actions.md:
```diff
@@ -0,0 +1,296 @@
+- No additional effects
+
+### WDL:
+        - turn_on:
```
```suggestion
```
        - turn_on:
```
```

SnowballAntrobus commented on docs/actions.md:
```diff
@@ -0,0 +1,296 @@
+- No additional effects
+
+### WDL:
+        - turn_off:
```
```suggestion
```
        - turn_off:
```
```

SnowballAntrobus commented on docs/actions.md:
```diff
@@ -0,0 +1,296 @@
+- Brief description of room will be printed
+
+### WDL:
+        - go:
```
```suggestion
```
        - go:
```
```

SnowballAntrobus commented on docs/actions.md:
```diff
@@ -0,0 +1,296 @@
+- Decrease player inventory space
+
+### WDL:
+        - take:
```
```suggestion
```
        - take:
```
```

SnowballAntrobus commented on docs/actions.md:
```diff
@@ -0,0 +1,296 @@
+- Put Item in room (changes room description and state)
+
+### WDL:
+        - drop:
```
```suggestion
```
        - drop:
```
```

SnowballAntrobus commented on docs/actions.md:
```diff
@@ -0,0 +1,296 @@
+- No additional effects
+
+### WDL:
+        - use:
```
```suggestion
```
        - use:
```
```

SnowballAntrobus commented on docs/actions.md:
```diff
@@ -0,0 +1,296 @@
+- No additional effects
+
+### WDL:
+        - use_on:
```
```suggestion
```
        - use_on:
```
```

SnowballAntrobus commented on docs/actions.md:
```diff
@@ -0,0 +1,296 @@
+- No additional effects
+
+### WDL:
+        - drink:
```
```suggestion
```
        - drink:
```
```

SnowballAntrobus commented on docs/actions.md:
```diff
@@ -0,0 +1,296 @@
+- No additional effects
+
+### WDL:
+        - eat:
```
```suggestion
```
        - eat:
```
```

SnowballAntrobus commented:
think i did the mark changes

mpcsh commented:
@SnowballAntrobus doesn't look like it. Click the "files" tab - there are still quite a few unresolved things (some that I opened, some that you opened).

SnowballAntrobus commented:
@mpcsh My bad the unresolved I opened were a holdover for when I tried to do all the changes from “GitHub suggestions”. They are now closed and were resolved in the previous commits. The unresolved that you opened that are still open are not in the spec doc so I didn’t think it was my responsibility. @MTUC08 would you like me to make those changes too?

lucyyutingli commented on src/action_management/include/action_structs.h:
```diff
@@ -0,0 +1,75 @@
+*/
+typedef struct {
+    enum actions act;   // e.g. CONSUME
+    char *c_name;  // e.g. "eat"
```
I believe we’ve changed the struct completely. All comments reflect the changes.

lucyyutingli commented on src/action_management/include/actionmanagement.h:
```diff
@@ -0,0 +1,198 @@
+ * - a: An action struct. Must point to action struct allocated by action_new
+ *
+ * Returns:
+ * - Always returns 0.
```
Changed it to void.

MTUC08 commented:
@SnowballAntrobus I should've done all the requested changes on the h/c files unless they were undone by the other merge.

yoonsjeong commented on src/action_management/include/actionmanagement.h:
```diff
@@ -0,0 +1,198 @@
+ * Returns:
+ * - a linked list of action_t structs of the same action kind
+ */
+list_t *get_actions_kind1();
```
we've since merged all these proto "get_actions" functions into a single "get_supported_actions" function

MTUC08 commented:
@mpcsh everything should be good to go now

mpcsh commented on src/action_management/include/action_structs.h:
```diff
@@ -0,0 +1,65 @@
+typedef struct {
+    action_type_t *act;
+    struct list_actions *next;
+} list_actions;
```
So this isn't a list of actions, this is a list of action types. The name of this struct should reflect that. Also, it should be suffixed by `_t`, so: `list_action_type_t`.

mpcsh commented on src/action_management/src/actionmanagement.c:
```diff
@@ -0,0 +1,197 @@
+
+
+/* See actionmanagement.h */
+action_type_t *action_new(char *c_name, enum action_kind kind)
```
These functions should be called `action_type_new`, `action_type_init`, etc. This is going to get very confusing if we aren't vigilant about naming.

mpcsh commented on src/action_management/src/actionmanagement.c:
```diff
@@ -0,0 +1,197 @@
+
+// KIND 1
+/* See actionmanagement.h */
+int action_item(game_t *g, action_type_t *a, item_t *i)
```
Might also want to rename these functions to `do_item_action`, `do_path_action`, etc.

mpcsh commented on src/action_management/src/actionmanagement.c:
```diff
@@ -0,0 +1,197 @@
+    case OPEN:
+    case CLOSE:
+    case TURN_ON:
+    case TURN_OFF: {
```
`case` blocks don't need braces, you can just do
```
case BLAH:
    blah();
    blah = blah;
    break;
case BLAH:
    ...
```

MTUC08 replied:
If I don't add the braces, the code breaks because I'm declaring functions within. How do I remedy this? @mpcsh

MTUC08 replied:
I just readied the braces is that ok?

mpcsh replied:
Hm, yeah that's fine. I didn't think that would break.

mpcsh commented on src/action_management/src/get_actions.c:
```diff
@@ -0,0 +1,81 @@
+
+
+static action_type_t valid_actions[] = {
+// KIND 1
```
Comments should be indented to the same level as the code they are commenting on.

mpcsh commented on src/action_management/src/get_actions.c:
```diff
@@ -0,0 +1,81 @@
+// KIND 1
+    {"open", ITEM},
+    {"close", ITEM},
+    {"push",ITEM},
```
Missing a space after the comma and before `ITEM`. Going to have to make this change in several places.

mpcsh commented on src/action_management/src/get_actions.c:
```diff
@@ -0,0 +1,81 @@
+static int INDEX_ITEM_ITEM = 12;
+
+
+list_actions *get_supported_actions(enum action_kind kind)
```
So this is... hacky. Why are we forcing CLI to ask for all the actions of a specific kind? Why can't we just have this function take no parameters and return a list of *all* the actions?

MTUC08 replied:
@yoonsjeong

ellychoi replied:
I've made the other changes but the team is still discussing this action. It'll be updated shortly!

yoonsjeong replied:
I'll fix it so that the function now returns just a full list of all supported actions.

yoonsjeong replied:
This has bee resolved in https://github.com/uchicago-cs/chiventure/pull/77/commits/48d3111dbe940c649ea874f383974902e2ce6881

mpcsh commented on src/action_management/include/actionmanagement.h:
```diff
@@ -18,7 +18,7 @@
  * - a pointer to a new action struct
  */
-action_type_t *action_new(char *c_name, enum action_kind kind);
+action_type_new *action_new(char *c_name, enum action_kind kind);
```
Uhhh... this isn't right...

ellychoi replied:
Hi, 
I'm slightly confused. I believe your comment was "These functions should be called action_type_new, action_type_init, etc. This is going to get very confusing if we aren't vigilant about naming". Should it be action_type_new_t?

mpcsh commented on src/action_management/Makefile:
```diff
@@ -0,0 +1,27 @@
+#Based off Makefile template from libobj which is from https://gist.github.com/xuhdev/1873316#
+
+CC = gcc
+SRCS = src/actionmanagement.c src/game.c
```
`get_actions.c` isn't being compiled! This explains my slack message - the call to `action_new` in that file shouldn't pass the compiler, but it's not being compiled.

mpcsh commented on src/action_management/src/get_actions.c:
```diff
@@ -0,0 +1,45 @@
+
+    for(i = 0; i < NUM_ACTIONS; i++) {
+        list_action_type_t *add = (list_action_type_t*)malloc(sizeof(list_action_type_t));
+        action_type_t *add_data = action_new(valid_actions[i].c_name,
```
`action_new` no longer exists

mpcsh commented on src/action_management/src/actionmanagement.c:
```diff
@@ -0,0 +1,194 @@
+        if (drop != SUCCESS) {
+            fprintf(stderr, "Object could not be removed from inventory.\n");
+            return FAILURE;
+        }
```
This needs a `break` because if `drop == SUCCESS`, then you don't return, and fall through to the next `case`.

mpcsh commented on src/action_management/include/actionmanagement.h:
```diff
@@ -0,0 +1,110 @@
+ * A function that initializes an action struct, filling its parameters.
+ *
+ * Parameters:
+ * - a: a pointer to an empty action_type_t allocated by action_new
```
`action_new` no longer exists

mpcsh commented on src/action_management/include/actionmanagement.h:
```diff
@@ -0,0 +1,110 @@
+ * A function that frees the resources associated with an action struct
+ *
+ * Parameters:
+ * - a: An action struct. Must point to action struct allocated by action_new
```
`action_new` no longer exists

mpcsh replied:
Also:
- An action *type* struct
- point to action *type* struct

mpcsh commented on src/action_management/include/actionmanagement.h:
```diff
@@ -0,0 +1,110 @@
+   =========================================================================== */
+
+/*
+ * Allocates a new action on the heap, calling action_init
```
Allocates a new action *type* on the heap

mpcsh commented on src/action_management/include/actionmanagement.h:
```diff
@@ -0,0 +1,110 @@
+ * - kind: an enumeration of the kind of action
+ *
+ * Returns:
+ * - a pointer to a new action struct
```
a pointer to a new action *type* struct

mpcsh commented on src/action_management/include/actionmanagement.h:
```diff
@@ -0,0 +1,110 @@
+
+
+/*
+ * A function that initializes an action struct, filling its parameters.
```
initializes an action *type* struct

mpcsh commented on src/action_management/include/actionmanagement.h:
```diff
@@ -0,0 +1,110 @@
+ *
+ * Returns:
+ * - 0 if success, 1 if an error occurs
+ * - an error message in stderr if the action struct is not initialized
```
if the action *type* struct is not initialized

mpcsh commented on src/action_management/include/actionmanagement.h:
```diff
@@ -0,0 +1,110 @@
+
+
+/*
+ * A function that frees the resources associated with an action struct
```
associated with an action *type* struct

mpcsh commented on src/action_management/include/actionmanagement.h:
```diff
@@ -0,0 +1,110 @@
+ *
+ * Parameters:
+ * - g: A game struct consisting of the game state
+ * - a: An action struct
```
action *type* struct

mpcsh commented on src/action_management/include/actionmanagement.h:
```diff
@@ -0,0 +1,110 @@
+ *
+ * Parameters:
+ * - g: A game struct consisting of the game state
+ * - a: An action struct
```
action *type* struct

mpcsh commented on src/action_management/include/actionmanagement.h:
```diff
@@ -0,0 +1,110 @@
+ *
+ * Parameters:
+ * - g: A game struct containing a field for the current player
+ * - a: An action struct
```
action *type* struct

mpcsh commented on src/action_management/src/actionmanagement.c:
```diff
@@ -0,0 +1,194 @@
+
+    int new_a = action_type_init(a, c_name, kind);
+    if (new_a != SUCCESS) {
+        fprintf(stderr, "Could not initialize this action");
```
this action *type*

mpcsh replied:
Also maybe print out `c_name` for debugging purposes?

lucyyutingli replied:
Fixed the action type part, I'll let one of the original writers approach the second one.

mpcsh commented on src/action_management/src/actionmanagement.c:
```diff
@@ -0,0 +1,194 @@
+    assert(a);
+    assert(i);
+    if (a->kind != ITEM) {
+        fprintf(stderr, "The action provided is not of the correct type.\n");
```
The action *type* provided is not of the correct ~~type~~ *kind*.

mpcsh commented on src/action_management/src/actionmanagement.c:
```diff
@@ -0,0 +1,194 @@
+    assert(g);
+    assert(a);
+    if (a->kind != PATH) {
+        fprintf(stderr, "The action provided is not of the correct type.\n");
```
The action *type* provided is not of the correct ~~type~~ *kind*.

mpcsh commented on src/action_management/src/actionmanagement.c:
```diff
@@ -0,0 +1,194 @@
+    assert(direct);
+    assert(indirect);
+    if (a->kind != ITEM_ITEM) {
+        fprintf(stderr, "The action provided is not of the correct type.\n");
```
The action *type* provided is not of the correct ~~type~~ *kind*.

lucyyutingli commented:
Closing Statements:

With this pull request we expanded on our specifications for different actions. They’ll be evolved as we continue on the project, and as we saw, changed as the requirements changed as well, resulting in us deleting some of the possible actions. We’ll be making functions to implement each of these, eventually focusing on making sure conditions are met for these actions to go through.

mpcsh commented:
Issue Score: ✓+

Comments:
Great work!

<!-- {"score": "check-plus"} -->
